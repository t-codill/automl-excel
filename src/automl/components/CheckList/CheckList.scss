.ms-List {
    &.checkList {
        border: 2px solid lightgray;
        padding: "2px";
    }
}
.ms-Grid {
    &.checkListItem {
        padding: 5px;
    }
}
