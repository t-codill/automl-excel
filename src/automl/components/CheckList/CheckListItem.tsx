import { Checkbox } from "office-ui-fabric-react";
import * as React from "react";

export interface ICheckListItemProps<T> {
    item: T;
    onChange?(item: T, checked?: boolean): void;
}

export class CheckListItem<T> extends React.Component<ICheckListItemProps<T>>{
    public render(): React.ReactNode {
        return <>
            <div className="ms-Grid checkListItem" dir="ltr">
                <div className="ms-Grid-col">
                    {/*key must be set to a unique value that changes when the item changes to force a re-render of child component. */}
                    <Checkbox className="listItemCheckbox" key={JSON.stringify(this.props.item)}
                        onChange={this.onCheckBoxChange}
                    />
                </div>
                <div className="ms-Grid-col">
                    {this.props.item}
                </div>
            </div>
        </>;
    }

    private readonly onCheckBoxChange = (_ev?: React.FormEvent<HTMLElement | HTMLInputElement>, checked?: boolean) => {
        if (this.props.onChange) {
            this.props.onChange(this.props.item, checked);
        }
    }
}
