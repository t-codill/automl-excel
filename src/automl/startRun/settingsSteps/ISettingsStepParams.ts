import { AzureMachineLearningJasmineManagementModels } from "@vienna/jasmine";
import { RunType } from "../../services/constants/RunType";

export interface ISettingsStepParams {
    jobType: RunType | undefined;
    column: string | undefined;
    metric: AzureMachineLearningJasmineManagementModels.PrimaryMetric | undefined;
    trainingJobTime: string | undefined;
    maxIteration: string | undefined;
    metricThreshold: string | undefined;
    maxConcurrentIterations: string | undefined;
    maxCores: string | undefined;
    preprocessing: boolean | undefined;
    blacklistAlgos: string[];
    crossValidationNumber: string | undefined;
    percentageValidation: string | undefined;
    timeSeriesColumn: string | undefined;
    maxHorizon: string | undefined;
    grainColumns: string[];
}
