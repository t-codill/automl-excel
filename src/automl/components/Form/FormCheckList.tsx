import * as React from "react";
import { CheckList, ICheckListProps } from "../CheckList/CheckList";
import { FormBaseInput } from "./FormBaseInput";

export class FormCheckList<
  TData extends { [key in TField]: string[] },
  TField extends keyof TData
  > extends FormBaseInput<TData, TField, string[], ICheckListProps>{
  public render(): React.ReactNode {
    return (
      <CheckList
        items={[]}
        {...this.props}
        onChange={this._onChange}
        values={this.state.value}
        errorMessage={this.state.errorMessage}
      />
    );
  }

  private readonly _onChange = (value: string[]): void => {
    this.setValue(value);
  }
}
