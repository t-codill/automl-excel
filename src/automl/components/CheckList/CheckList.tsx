import { isEqual } from "lodash";
import { List } from "office-ui-fabric-react";
import * as React from "react";
import { ValidationError } from "../Form/ValidationError";
import { CheckListItem } from "./CheckListItem";

import "./CheckList.scss";

export interface ICheckListProps {
    id?: string;
    placeholder?: string;
    items: string[];
    maxWidth?: string;
    values?: string[];
    errorMessage?: string;
    onChange?(items: string[] | undefined): void;
}

interface ICheckListState {
    selectedItems: string[];
}

export class CheckList extends React.Component<ICheckListProps, ICheckListState> {
    public constructor(props: ICheckListProps) {
        super(props);
        this.state = {
            selectedItems: this.props.values || []
        };
    }

    public componentDidUpdate(prevProps: ICheckListProps, _prevState: ICheckListState): void {
        if (!isEqual(prevProps.items, this.props.items)) {
            this.setState({ selectedItems: [] });
            if (this.props.onChange) {
                this.props.onChange([]);
            }
        }
    }

    public render(): React.ReactNode {
        return <>
            <List
                id={this.props.id}
                items={this.props.items}
                onRenderCell={this.onRenderCell}
                className="checkList"
                defaultValue={this.state.selectedItems}
                placeholder={this.props.placeholder}
            />
            <ValidationError text={this.props.errorMessage} />
        </>;
    }

    private readonly onSelectionUpdate = (item: string, checked?: boolean) => {
        const selectedItems = this.state.selectedItems;
        const itemIndex = selectedItems.findIndex((other) => other === item);

        if (checked && itemIndex < 0) {
            selectedItems.push(item);
        } else if (!checked) {
            selectedItems.splice(itemIndex, 1);
        }

        this.setState({ selectedItems });

        if (this.props.onChange) {
            this.props.onChange(selectedItems);
        }
    }

    private readonly onRenderCell = (item?: string): React.ReactNode => {
        if (item === undefined) {
            return <></>;
        }
        return <CheckListItem item={item} onChange={this.onSelectionUpdate} />;
    }
}
