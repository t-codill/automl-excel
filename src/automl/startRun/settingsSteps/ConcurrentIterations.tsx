import { AzureMachineLearningWorkspacesModels } from "@azure/arm-machinelearningservices";
import * as React from "react";
import { FormTextInput } from "../../components/Form/FormTextInput";
import { Validators } from "../../components/Form/Validators";
import { LabelWithCallout } from "../../components/Label/LabelWithCallout";
import { ISettingsStepParams } from "./ISettingsStepParams";

interface IConcurrentIterationInputProps {
    compute: AzureMachineLearningWorkspacesModels.ComputeUnion | undefined;
}

const concurrentIterations: React.FunctionComponent<IConcurrentIterationInputProps> = (props) => {
    return <div className="ms-Grid" dir="ltr">
        <div className="ms-Grid-col ms-sm12 ms-md6 ms-lg6 ms-xl6">
            <LabelWithCallout required={false} htmlFor={"maxConcurrentIterInput"} labelText="Max concurrent iterations">
                Maximum number of iterations that would be executed in parallel. This should be less than the number of cores on the compute target.
            </LabelWithCallout>
        </div>
        <div className="ms-Grid-col ms-sm12 ms-md6 ms-lg6 ms-xl6">
            <FormTextInput<ISettingsStepParams, "maxConcurrentIterations">
                field="maxConcurrentIterations"
                placeholder="Concurrent Iterations"
                ariaLabel="Please enter the max number of concurrent iterations here"
                type="number"
                validators={[
                    Validators.isInteger("Max concurrent iteration should be an integer"),
                ]}
                defaultFormValue={`${
                    props.compute
                        && props.compute.computeType === "AmlCompute"
                        && props.compute.properties
                        && props.compute.properties.scaleSettings
                        ? props.compute.properties.scaleSettings.maxNodeCount : 1}`}
            />
        </div>
    </div>;
};

export { concurrentIterations as ConcurrentIterations };
