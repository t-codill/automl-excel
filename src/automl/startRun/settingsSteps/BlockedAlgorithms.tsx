import * as React from "react";
import { FormCheckList } from "../../components/Form/FormCheckList";
import { LabelWithCallout } from "../../components/Label/LabelWithCallout";
import { Algorithms } from "../../services/constants/Algorithms";
import { RunType } from "../../services/constants/RunType";
import { ISettingsStepParams } from "./ISettingsStepParams";

export interface IBlockedAlgorithmsProps {
    jobType: RunType;
}

interface IBlockedAlgorithmsState {
    algorithms: string[];
}

export class BlockedAlgorithms extends React.Component<IBlockedAlgorithmsProps, IBlockedAlgorithmsState> {

    constructor(props: IBlockedAlgorithmsProps) {
        super(props);
        this.state = {
            algorithms: this.getAlgorithms()
        };
    }

    public componentDidUpdate(prevProps: IBlockedAlgorithmsProps): void {
        if (prevProps.jobType !== this.props.jobType) {
            this.setState({ algorithms: this.getAlgorithms() });
        }
    }

    public render(): React.ReactNode {
        return <>
            <LabelWithCallout required={true} htmlFor={undefined} labelText="Blocked Algorithms">
                Select algorithms to exclude
            </LabelWithCallout>
            <FormCheckList<ISettingsStepParams, "blacklistAlgos">
                field="blacklistAlgos"
                items={this.state.algorithms}
                validators={[this.validateSelection]}
            />
        </>;
    }

    private readonly validateSelection = (value: string[] | undefined) => {
        if (value && value.length === this.state.algorithms.length) {
            return "Must allow at least one algorithm";
        }
        return undefined;
    }

    private readonly getAlgorithms = (): string[] => {
        switch (this.props.jobType) {
            case "classification":
                return Algorithms.classification;
            case "regression":
                return Algorithms.regression;
            default:
                return Algorithms.regression;
        }
    }
}
