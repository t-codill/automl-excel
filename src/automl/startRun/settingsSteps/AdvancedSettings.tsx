import { AzureMachineLearningWorkspacesModels } from "@azure/arm-machinelearningservices";
import * as React from "react";
import { RunType } from "../../services/constants/RunType";
import { BlockedAlgorithms } from "./BlockedAlgorithms";
import { ConcurrentStep } from "./ConcurrentStep";
import { ExitCriteria } from "./ExitCriteria";
import { MetricSelector } from "./MetricSelector";
import { Preprocessing } from "./Preprocessing";
import { Validation } from "./Validation";

export class AdvancedSettings extends React.Component<{
    jobType: RunType;
    compute: AzureMachineLearningWorkspacesModels.ComputeUnion | undefined;
    primaryMetric: string;
}>{
    public render(): React.ReactNode {
        return <>
            <div className="ms-md12 ms-lg8 ms-xl6 ms-xxl4">
                <MetricSelector jobType={this.props.jobType} />
            </div>
            <div className="ms-md12 ms-lg8 ms-xl6 ms-xxl4">
                <ExitCriteria primaryMetric={this.props.primaryMetric} />
            </div>
            <div className="ms-md12 ms-lg8 ms-xl6 ms-xxl4">
                <Preprocessing
                    jobType={this.props.jobType} />
            </div>
            <div className="ms-md12 ms-lg8 ms-xl6 ms-xxl4">
                <Validation jobType={this.props.jobType} />
            </div>
            <div className="ms-md12 ms-lg8 ms-xl6 ms-xxl4">
                <ConcurrentStep compute={this.props.compute} />
            </div>
            <div className="ms-md12 ms-lg8 ms-xl6 ms-xxl4">
                <BlockedAlgorithms jobType={this.props.jobType} />
            </div>
        </>;
    }

}
