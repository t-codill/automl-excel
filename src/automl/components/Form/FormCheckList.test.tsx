import { shallow, ShallowWrapper } from "enzyme";
import * as React from "react";
import { ISettingsStepParams } from "../../startRun/settingsSteps/ISettingsStepParams";
import { CheckList, ICheckListProps } from "../CheckList/CheckList";
import { FormCheckList } from "./FormCheckList";

jest.mock("./FormBaseInput");

describe("FormCheckList", () => {
  let wrapper: ShallowWrapper<{}, { value: string | undefined }>;
  let formCheckListProps: ICheckListProps;
  beforeEach(() => {
    wrapper = shallow(
      <FormCheckList<ISettingsStepParams, "blacklistAlgos">
        field="blacklistAlgos"
        items={[""]}
      />);
    formCheckListProps = wrapper.find(CheckList)
      .props();
  });

  it("should render", () => {
    expect(wrapper)
      .toMatchSnapshot();
  });

  it("should allow entering free text", () => {
    if (formCheckListProps.onChange) {
      formCheckListProps.onChange(["b"]);
    }
    expect(wrapper.state("value"))
      .toEqual(["b"]);
  });
});
