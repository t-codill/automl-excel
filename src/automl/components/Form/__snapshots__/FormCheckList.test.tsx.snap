// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormCheckList should render 1`] = `
<CheckList
  field="blacklistAlgos"
  items={
    Array [
      "",
    ]
  }
  onChange={[Function]}
/>
`;
